src/main/java/com/whiteclarkegroup/liquibaselinter/report/AbstractReporter.java

```diff
@@ -0,0 +1,123 @@
+package com.whiteclarkegroup.liquibaselinter.report;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.collect.ImmutableSet;
+import com.whiteclarkegroup.liquibaselinter.report.ReporterConfig.BaseBuilder;
+import liquibase.exception.UnexpectedLiquibaseException;
+import org.springframework.core.GenericTypeResolver;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.whiteclarkegroup.liquibaselinter.report.ReportItem.ReportItemType.ERROR;
+import static com.whiteclarkegroup.liquibaselinter.report.ReportItem.ReportItemType.IGNORED;
+
+public abstract class AbstractReporter implements Reporter {
+    private final ReporterConfig config;
+    private final String path;
+    private final Set<ReportItem.ReportItemType> filter;
+
+    protected AbstractReporter(ReporterConfig config, String defaultPathExtension) {
+        this.config = config;
+        this.path = Optional.ofNullable(config.getPath()).orElse("./target/lqlint-report." + defaultPathExtension);
+        this.filter = Optional.ofNullable(config.getFilter()).map(ImmutableSet::copyOf).orElse(ImmutableSet.of(ERROR, IGNORED));
+    }
+
+    @Override
+    public ReporterConfig getConfiguration() {
+        return config;
+    }
+
+    @Override
+    public void processReport(Report report) {
+        final List<ReportItem> filteredItems = report.getItems().stream()
+            .filter(item -> filter.contains(item.getType()))
+            .collect(Collectors.toList());
+        process(report, filteredItems);
+    }
+
+    protected void process(Report report, List<ReportItem> items) {
+        try (FileWriter writer = new FileWriter(path)) {
+            printReport(new PrintWriter(writer), report, items);
+            writer.flush();
+        } catch (IOException e) {
+            throw new UnexpectedLiquibaseException("Unable to write text report", e);
+        }
+    }
+
+    protected abstract void printReport(PrintWriter output, Report report, List<ReportItem> items);
+
+    protected static long countDisabledRules(Report report) {
+        return report.getConfig().getRules().values().stream().filter(rule -> !rule.isEnabled()).count();
+    }
+
+    public abstract static class BaseFactory<R extends Reporter, C extends ReporterConfig> implements Reporter.Factory<R, C> {
```

Comment:
Is this complexity mainly to work around the fact that we want immutable reporters if possible, and with raw SPI that is not possible because we must have a default constructor. If so, do we think that complexity is worth it, vs just having a simple `configure` method in the reporter, similar to what we have for the rules?

Reply:
In general, I do prefer immutability as it prevents many multi-threading issues or attack vectors. I know neither of those are real concerns right now in the `liquibase-linter`, but I could certainly see benefit in the rules being immutable for the purpose of parallelizing their evaluation of changes. Would it be beneficial to speed up linting on large change sets?

The concrete factories do have a no arg constructor, so they meet that SPI standard. How the reporter is created from the factory is not dictated in any way by the interfaces. If someone were to create a class that implemented `Reporter` and provided a `setConfiguration` method that its corresponding factory called, it would be free to do so. But it would not be the expectation that the `liquibase-linter` would be reconfiguring these reporters constantly. 

Yes, it adds a little bit more complexity vs the rules, but when I started really diving into the `liquibase-linter`, the fact that rules are constantly reconfigured caused be to go "Huh?". I didn't want to create that same "huh?" for the reporters. And I do have code that I plan to submit that makes rules follow the same pattern. I'm not expecting to debate that right now here on this PR. I'm just stating that if your concerns are for consistency, I do plan to follow through and create consistency.

Reply:
Also, this added complexity enables custom, rule specific configuration to be added to the `lqlint.json` file. In essence, this factory links a reporter name with a JSON schema so the config loader can parse the rule specific configuration. In a future PR, I would like to enable the same custom configuration options as I did here. 

Have I convinced you of this pattern?